_rules/heading-descriptive-b49b2e.md

```diff
@@ -29,13 +29,16 @@ This rule applies to any element with the [semantic role][] of heading that is e
 
 ## Expectation
 
-Each target element describes the topic or purpose of its [section of the content][].
+Each target element describes the topic or purpose of the [programmatic section of content][] it is [associated][] with.
 
 **Note:** Headings do not need to be lengthy. A word, or even a single character, may suffice.
 
 ## Assumptions
 
-This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content), and sufficiently understood.
+- This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content), and sufficiently understood.
+- This rule assumes that the [semantic roles][semantic role] (including the level of headings) are used according to their semantics.
```

Comment:
I'm not comfortable with this assumption. I understand this is the way to make a programatically determined section work, but I'm not so sure that that's the right way to go about this.  I don't think we have any reason to assume this is true. WCAG doesn't require it, nor is there much else that would enforce it.

In my opinion, what needs to happen to fix 'section of the content" in this rule is to make that definition explicitly about how the content is presented. Define a section of the content based on visual cues, like differences in background, border, whitespace, and based on what is the next heading with the same presentation.

Reply:
While I agree that this assumption is not the way to go, I don't see how a definition based on visual cues would work either. We can never really describe the maried ways in which this could be made visually apparent.

I think the definition should be based on what is the next heading with the same presentation... and then state that it can also be defined by visual cues (with examples) but that this is inherently subjective and will differ from audit to audit.

Reply:
I am aware that WCAG considers that styling is enough to make a heading. (and that if one uses only `font-size: large` to mark headings, this is goings to be a problem for "Info & Relationship").

If we base sections of content purely in terms of visual cues, this is however going to be completely irrelevant for screenreaders who do not have access to these visual cues.

The rule is specifically targeting `role="heading"` hinting that it is not about visual cues only (maybe that needs to change also).

Reply:
I strongly disagree with you Brian. Take this example:
https://codepen.io/wilcofiers/pen/abOvWRB

It is visually clear that the bottom paragraph is not intended to be part of the h2 subsection. This kind of a "card" design pattern is very common. As for why I think we can make this rule about programatic headings, but not about programatic sections, that's because heading markup is required by WCAG. I don't think WCAG requires the same for sections, nor should it, because there's no commonly agreed upon definition of what that even is. That's why you couldn't just use an existing definition for this proposal.

Reply:
Yes, my proposal would put the bottom paragraph as part of the last `h2` section. Which is not what is perceived by sighted users.

But, wouldn't an AT/screen reader do the same (put the bottom paragraph in the last `h2`)? Because it can't detect the visual border marking the cards. I guess this fails Info&Relationship in any case…

Reply:
> But, wouldn't an AT/screen reader do the same (put the bottom paragraph in the last h2 section

I don't think they would. As far as I know, AT don't try to identify implicit sections. They might communicate something for the `section` element (I have no idea TBH, but they might), but I doubt they would do something with "implicit sections". They just list the headings and what levels they have. That structure is entirely separate from the landmark / region structure. It doesn't get combined in some way like you're trying to do here. These just two different structures that AT can communicate.

The real question is; should AT users be aware that (in my example) that bottom paragraph doesn't belong to the h2? That's a 1.3.1 question. If 1.3.1 required this, I would have an issue with this assumption. But I don't think it does. And yes, that means AT users lose out on some information that is visually evident, but isn't programatic. That is also true for not using landmarks, and most WCAG auditors don't consider that a requirement of 1.3.1 either.

Reply:
Hum, I've played a bit with your CodePen and NVDA and I think I understand what you mean (I have little experience with AT usage…)

AT don't really care where Sections end. Only where they begin. AT provide a list of heading and a way to jump to a heading but doesn't really have a "read until end of section (next heading)" thingie. Similarly, AT provide a list of landmarks and a way to jump to it, but not a "read content of that landmark thingie".

The heading and landmark structures are kept separate by AT (unlike the outline algorithm) and maybe we should keep it separate also for this rule.

---
A problem with this rule is that it is very much concerned by where the section ends…
Building on your codepen:
```
<h1>Fruits</h1>
<div class="row">
  <div class="cell"><h2>Apples</h2>Apples are cool.</div>
  <div class="cell"><h2>Pears</h2>Pears are cool.</div>
</div>
Other fruits also exist.
```
The result on this rule in this example pretty much depends on where the last sentence is. If it is part of the last `h2` (as pure heading structure tell), then this fails the rule because "Pears" is not descriptive of "Pears are cool. Other fruits also exist". But if it is not part of the last `h2`, then this pass the rule…
(same question if the `div` are replaced by `section` and we only look at the heading structure to determine the sections of content…)

=> Would you say that it fails 1.3.1? (I'd say yes because the fact that the last sentence is not in any `h2` section is not communicated programmatically)
=> Would you say that this example fails 2.4.6? (It seems to fail it for AT users but not for sighted users. Obviously related to the fact that it is failing 1.3.1 so different users perceive different structure in the page. Does that mean we should split the rule? Or have an assumption that the page passe 1.3.1?)
=> What should we (and the "heading is descriptive" rule) do about this example?

---
What about having the implicit/explicit sections (more or less as defined here) and an assumption  in the line of "if both are used, they do not overlap in weird ways" rather than "elements are used according to semantics"?
Then something like
```html
<section>
  <h1>Fruits</h1>
  <section><h2>Apples</h2>Apples are cool.</section>
  <section><h2>Pears</h2>Pears are cool.</section>
Other fruits also exist.
</section>
```
would fail the assumption (because the "Pears" implicit section overlaps in weird ways with the explicit sections) and we don't care about what we output for it…
(the same example with `div` would still be poorly handled, however)

And maybe this rule should target only the implicit sections…

Reply:
> 
> 
> I strongly disagree with you Brian. Take this example:
> https://codepen.io/wilcofiers/pen/abOvWRB
> 
> It is visually clear that the bottom paragraph is not intended to be part of the h2 subsection. This kind of a "card" design pattern is very common. As for why I think we can make this rule about programatic headings, but not about programatic sections, that's because heading markup is required by WCAG. I don't think WCAG requires the same for sections, nor should it, because there's no commonly agreed upon definition of what that even is. That's why you couldn't just use an existing definition for this proposal.

I don't see how that is contradicting anything I said. I agree with this message.